porch/repository/pkg/git/commit.go

```diff
@@ -0,0 +1,348 @@
+	} else {
+		parentCommit, err := object.GetCommit(storer, parent)
+		if err != nil {
+			return nil, fmt.Errorf("cannot resolve parent commit hash %s to commit: %w", parent, err)
+		}
+		t, err := parentCommit.Tree()
+		if err != nil {
+			return nil, fmt.Errorf("cannot resolve parent commit's (%s) tree (%s) to tree object: %w", parent, parentCommit.TreeHash, err)
+		}
+		root = t
+	}
+
+	trees, err := initializeTrees(storer, root, packagePath, packageTree)
+	if err != nil {
+		return nil, err
+	}
+
+	ch := &commitHelper{
+		storer: storer,
+		trees:  trees,
+		parent: parent,
+	}
+
+	return ch, nil
+}
+
+// Initializes ancestor trees of the package, reading them from the storer.
+// If packageTree hash is provided, it will be used as the package's initial tree. Otherwise, new tree will be used
+// (effectively replacing the package with an empty one).
+func initializeTrees(storer storage.Storer, root *object.Tree, packagePath string,
+	packageTreeHash plumbing.Hash) (map[string]*object.Tree, error) {
+
+	trees := map[string]*object.Tree{
+		"": root,
+	}
+
+	parts := strings.Split(packagePath, "/")
+	if len(parts) == 0 {
+		// empty package path is invalid
+		return nil, fmt.Errorf("invalid package path: %q", packagePath)
+	}
+
+	// Load all ancestor trees
+	parent := root
+	for i, max := 0, len(parts)-1; i < max; i++ {
+		name := parts[i]
+		path := strings.Join(parts[0:i+1], "/")
+
+		var current *object.Tree
+		switch existing := findEntry(parent, name); {
+		case existing == nil:
+			// Create new empty tree for this ancestor.
+			current = &object.Tree{}
+
+		case existing.Mode == filemode.Dir:
+			// Existing entry is a tree. use it
+			hash := existing.Hash
+			curr, err := object.GetTree(storer, hash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot read existing tree %s; root %q, path %q", hash, root.Hash, path)
+			}
+			current = curr
+
+		default:
+			// Existing entry is not a tree. Error.
+			return nil, fmt.Errorf("path %q is %s, not a directory in tree %s, root %q", path, existing.Mode, existing.Hash, root.Hash)
+		}
+
+		// Set tree in the parent
+		setOrAddDirEntry(parent, name)
+
+		trees[strings.Join(parts[0:i+1], "/")] = current
+		parent = current
+	}
+
+	// Initialize the package tree.
+	lastPart := parts[len(parts)-1]
+	if !packageTreeHash.IsZero() {
+		// Initialize with the supplied package tree.
+		packageTree, err := object.GetTree(storer, packageTreeHash)
+		if err != nil {
+			return nil, fmt.Errorf("cannot find existing package tree %s for package %q", packageTreeHash, packagePath)
+		}
+		trees[packagePath] = packageTree
+		setOrAddDirEntry(parent, lastPart)
+	}
+
+	return trees, nil
+}
+
+// Returns index of the entry if found (by name); -1 if not found
+func findEntry(tree *object.Tree, name string) *object.TreeEntry {
+	for i := range tree.Entries {
+		e := &tree.Entries[i]
+		if e.Name == name {
+			return e
+		}
+	}
+	return nil
+}
+
+func setOrAddDirEntry(tree *object.Tree, name string) {
```

Comment:
If this gets more complicated, we could consider a new type `MutableGitTree` or similar.  It would likely just have `[]object.TreeEntry`, but it could at least encapsulate the logic!